elements/lisk-p2p/src/peer_server/peer_server.ts

```diff
@@ -0,0 +1,460 @@
+			clearTimeout(this._invalidMessageInterval);
+		}
+
+		await this._stopWSServer();
+		await this._stopHTTPServer();
+	}
+
+	public async start(): Promise<void> {
+		this._invalidMessageInterval = setInterval(() => {
+			this._invalidMessageCounter = new Map();
+		}, DEFAULT_RATE_CALCULATION_INTERVAL);
+
+		this._scServer.addMiddleware(
+			this._scServer.MIDDLEWARE_HANDSHAKE_WS,
+			(
+				req: http.IncomingMessage,
+				next: SCServer.nextMiddlewareFunction,
+			): void => {
+				// Decline connections from banned IPs
+				if (this._peerBook.bannedIPs.has(req.socket.remoteAddress as string)) {
+					next(
+						new PeerInboundHandshakeError(
+							FORBIDDEN_CONNECTION_REASON,
+							FORBIDDEN_CONNECTION,
+							req.socket.remoteAddress as string,
+						),
+					);
+
+					return;
+				}
+
+				next();
+
+				return;
+			},
+		);
+
+		this._scServer.on('handshake', (socket: SCServerSocket): void => {
+			if (this._peerBook.bannedIPs.has(socket.remoteAddress)) {
+				if ((socket as any).socket) {
+					(socket as any).socket.terminate();
+				}
+			}
+
+			this._bindInvalidControlFrameEvents(socket);
+		});
+
+		// Handle incoming invalid payload
+		(this._scServer as any).wsServer.on('connection', (ws: any, req: any) => {
+			this._handleIncomingPayload(ws, req);
+		});
+
+		this._scServer.on('connection', (socket: SCServerSocket): void => {
+			this._handleIncomingConnection(socket);
+		});
+
+		this._httpServer.listen(
+			this._nodeInfo.wsPort,
+			this._hostIp || DEFAULT_NODE_HOST_IP,
+		);
+
+		if (this._scServer.isReady) {
+			return;
+		}
+
+		return new Promise<void>(resolve => {
+			this._scServer.once('ready', () => {
+				resolve();
+			});
+		});
+	}
+
+	private _terminateIncomingSocket(
+		socket: SCServerSocket,
+		error: Error | string,
+		addToBannedPeers?: boolean,
+	): void {
+		if ((socket as any).socket) {
+			(socket as any).socket.terminate();
+		}
+		// Re-emit the message to allow it to bubble up the class hierarchy.
+		this.emit(EVENT_INBOUND_SOCKET_ERROR, error);
+
+		// If the socket needs to be permanently banned
+		if (addToBannedPeers) {
+			const peerId = `${socket.remoteAddress}:${socket.remotePort}`;
+
+			this.emit(EVENT_BAN_PEER, peerId);
+		}
+	}
+
+	private _disconnectSocketDueToFailedHandshake(
+		socket: SCServerSocket,
+		statusCode: number,
+		closeReason: string,
+	): void {
+		socket.disconnect(statusCode, closeReason);
+
+		this.emit(
+			EVENT_FAILED_TO_ADD_INBOUND_PEER,
+			new PeerInboundHandshakeError(
+				closeReason,
+				statusCode,
+				socket.remoteAddress,
+				socket.request.url,
+			),
+		);
+	}
+
+	private _handleIncomingConnection(socket: SCServerSocket): void {
+		if (!socket.request.url) {
+			this._disconnectSocketDueToFailedHandshake(
+				socket,
+				INVALID_CONNECTION_URL_CODE,
+				INVALID_CONNECTION_URL_REASON,
+			);
+
+			return;
+		}
+		const queryObject = url.parse(socket.request.url, true).query;
+
+		if (queryObject.nonce === this._nodeInfo.nonce) {
```

Comment:
check for the query? `if (queryObject && query.nonce === this._nodeInfo.nonce) {`